File: python/0040组合总和Ⅱ.py



def combinationSum2(candidates: list, target: int) -> list:
    res=[]
    temp=[]
    
    def _backtraking(start:int,s=int,candidates=candidates,target=target):
        used=set()

        if(s>target):
            return
        if(s==target):
            res.append(temp.copy())
            return
        
        for i in range(start,len(candidates)):
            val=candidates[i]
            if(val in used):
                continue
            temp.append(val)
            used.add(val)
            _backtraking(i+1,s+val)
            temp.pop()

    candidates.sort()
    _backtraking(0,0)
    return res

print(combinationSum2([10,1,2,7,6,1,5],8))

File: cpp/0040组合总和Ⅱ.cpp
#include<vector>
#include<unordered_set>
#include<algorithm>

using namespace std;

vector<vector<int>> res;
vector<int> temp;
    
void backTracking(int start, vector<int> &candidates, int target, int sum)
{
    unordered_set<int> set;
    if (sum == target)
    {
        res.push_back(temp);
        return;
    }

    for (int i = start; i < candidates.size() && sum <= target; ++i)
    {
        int val = candidates[i];
        if (set.find(val)!=set.end())
            continue;
        temp.push_back(val);
        set.insert(val);
        backTracking(i + 1, candidates, target, sum + val);
        temp.pop_back();
    }
}

vector<vector<int>> combinationSum2(vector<int> &candidates, int target)
{
    sort(candidates.begin(),candidates.end());
    backTracking(0, candidates, target, 0);
    return res;
}